lib/templates/lab2/lab2.dart

```diff
@@ -1,19 +1,172 @@
+      Book('Book2', 'Author2', 42),
+      Book('Book3', 'Author3', 37)
+    ];
+    List<Text> details = [];
+    for (final book in books) {
+      details.add(Text(book.bookDetails()));
+    }
+    return Column(
+      children: details,
+    );
+  }
+
+  Widget ex2() {
+    List<FictionBook> books = [
+      FictionBook('Book1', 'Author1', 13, 'Thriller'),
+      FictionBook('Book2', 'Author2', 42, 'Comedy'),
+      FictionBook('Book3', 'Author3', 37, 'Detective')
+    ];
+    List<Text> details = [];
+    for (final book in books) {
+      details.add(Text(book.bookDetails()));
+    }
+    return Column(
+      children: details,
+    );
+  }
+
+  Widget ex3() {
+    List<Book> books = [
+      Book('Book1', 'Author1', 13),
+      Book('Book2', 'Author2', 42),
+      Book('Book3', 'Author3', 37)
+    ];
+    List<Text> details = [];
+    for (final book in books) {
+      details.add(Text(book.toJson().toString()));
+    }
+    return Column(
+      children: details,
+    );
+  }
+
+  Widget ex4() {
+    return Column(
+      children: [
+        Text(
+            'length: 15, width 22, area: ${calculateArea(width: 22, length: 15)}'),
+        Text(
+            'radius: 37, area: ${calculateArea(length: 37, width: 0, shape: 'circle')}')
+      ],
+    );
+  }
+
+  Widget ex5() {
+    List<User> users = [
+      User('Book1', email: 'Author1', age: 13),
+      User('Book2', age: 42),
+      User('Book3', email: 'Author3')
+    ];
+    List<Text> details = [];
+    for (final book in users) {
+      details.add(Text(book.printUserDetails()));
+    }
+    return Column(
+      children: details,
+    );
+  }
 }
 
-class FictionBook {
-  // your code here
+class Book with JsonSerializable {
+  String title;
+  String author;
+  int numOfPages;
+
+  Book(this.title, this.author, this.numOfPages);
+
+  String bookDetails() {
+    return "$author - $title ($numOfPages)";
+  }
+}
+
+class FictionBook extends Book {
+  String genre;
+
+  FictionBook(super.title, super.author, super.numOfPages, this.genre);
+
+  @override
+  String bookDetails() {
+    return "$author - $title <$genre> ($numOfPages)";
+  }
 }
 
 mixin JsonSerializable {
-  // your code here
+  Map<String, dynamic> toJson() {
+    Map<String, dynamic> json = {};
+    switch (runtimeType) {
+      case const (Book):
+        Book book = this as Book;
+        json['title'] = book.title;
+        json['author'] = book.author;
+        json['numOfPages'] = book.numOfPages;
+        return json;
+    }
+    return json;
+  }
 }
 
-double calculateArea(){
-  // your code here
+double calculateArea(
+    {required double width,
+    required double length,
+    String shape = 'rectangle'}) {
+  switch (shape) {
+    case 'circle':
+      return 3.14 * length * length;
+    default:
+      return length * width;
+  }
 }
 
-class User{
-  // your code here
+class User {
+  String name;
+  int? age;
+  String? email;
+
+  User(this.name, {this.age, this.email});
```

Comment:
name has to be required

Reply:
Isn't it required by default when it's unnamed?

Reply:
Yes it is, but to make it also clear in code it is better to use named parameters, it is just suggestion, not count for grade